tests/kokkos-based/overwriting_matrix_matrix_product.cpp

```diff
@@ -0,0 +1,145 @@
+  gemm_gold_solution(A, B, C_gold);
+
+  stdla::matrix_product(KokkosKernelsSTD::kokkos_exec<>(), A, B, C);
+
+  // after kernel, A,B should be unchanged, C should be equal to C_gold.
+  // note that for A we need to visit all elements rowwise
+  // since that is how we stored above the preKernel values
+
+  if constexpr(std::is_same_v<value_type, float>){
+    // check A
+    std::size_t count=0;
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent1; ++j){
+	EXPECT_FLOAT_EQ(A(i,j), A_preKernel[count++]);
+      }
+    }
+
+    // check B
+    count=0;
+    for (std::size_t i=0; i<extent1; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_FLOAT_EQ(B(i,j), B_preKernel[count++]);
+      }
+    }
+
+    // check C
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_NEAR(C(i,j), C_gold(i,j), 1e-3);
+      }
+    }
+  }
+
+  else if constexpr(std::is_same_v<value_type, double>){
+    // check A
+    std::size_t count=0;
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent1; ++j){
+	EXPECT_DOUBLE_EQ(A(i,j), A_preKernel[count++]);
+      }
+    }
+
+    // check B
+    count=0;
+    for (std::size_t i=0; i<extent1; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_DOUBLE_EQ(B(i,j), B_preKernel[count++]);
+      }
+    }
+
+    // check C
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_NEAR(C(i,j), C_gold(i,j), 1e-9);
+      }
+    }
+  }
+
+  else if constexpr(std::is_same_v<value_type, std::complex<double>>){
+    // check A
+    std::size_t count=0;
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent1; ++j){
+	EXPECT_DOUBLE_EQ(A(i,j).real(), A_preKernel[count].real());
+	EXPECT_DOUBLE_EQ(A(i,j).imag(), A_preKernel[count++].imag());
+      }
+    }
+
+    // check B
+    count=0;
+    for (std::size_t i=0; i<extent1; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_DOUBLE_EQ(B(i,j).real(), B_preKernel[count].real());
+	EXPECT_DOUBLE_EQ(B(i,j).imag(), B_preKernel[count++].imag());
+      }
+    }
+
+    // check C
+    for (std::size_t i=0; i<extent0; ++i){
+      for (std::size_t j=0; j<extent2; ++j){
+	EXPECT_NEAR(C(i,j).real(), C_gold(i,j).real(), 1e-9);
+	EXPECT_NEAR(C(i,j).imag(), C_gold(i,j).imag(), 1e-9);
+      }
+    }
+  }
+}
+}//end anonym namespace
+
+TEST_F(blas2_signed_float_fixture, kokkos_overwriting_matrix_matrix_product)
+{
+  kokkos_blas_overwriting_gemm_impl(A_e0e1, B_e1e2, C_e0e2);
+}
+
+TEST_F(blas2_signed_double_fixture, kokkos_overwriting_matrix_vector_product)
+{
+  kokkos_blas_overwriting_gemm_impl(A_e0e1, B_e1e2, C_e0e2);
+}
+
+TEST_F(blas2_signed_complex_double_fixture, kokkos_overwriting_matrix_vector_product)
+{
+  using kc_t = Kokkos::complex<double>;
+  using stdc_t = value_type;
+  if constexpr (alignof(value_type) == alignof(kc_t)){
+    kokkos_blas_overwriting_gemm_impl(A_e0e1, B_e1e2, C_e0e2);
+  }
+}
```

Comment:
Would you consider adding some tests that exercise `conjugated`, `scaled`, and `transposed`?  That's especially important for the Kokkos case, where the back-end needs to extract information from those.  Another important test is `scaled` with a nested accessor that's not `default_accessor`.  It's important that this get sent to the fall-back implementation.

[Edit] I see that the code only includes specializations for `default_accessor`, so perhaps this isn't an issue yet.

Reply:
Yes I was planning to add those too! Thanks for bringing it up